engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/ReplayStateMachine.java

```diff
@@ -230,11 +229,19 @@ private void replayEvent(final LoggedEvent currentEvent) {
 
   /**
    * Ends the replay and sets some important properties, especially completes the replay future with
-   * the last source event, which has caused the last applied event.
+   * the last processing positions.
    */
   private void onRecordsReplayed() {
-    LOG.info(LOG_STMT_REPLAY_FINISHED, lastReadRecordPosition);
-    recoveryFuture.complete(lastSourceEventPosition);
+    final var lastProcessingPositions =
+        new LastProcessingPositions(
+            // the last source pos. is the last processed command by the processing state machine
+            lastSourceEventPosition,
+            // the last read position on replay, will be used as the last written for
+            // the processing state machine
+            lastReadRecordPosition);
```

Comment:
❓  Why does the lastReadRecordPosition match the last written position? Does the replay state machine read all the way to the end of the log? If so, please expand the comment to describe why it's possible to use it here.

Reply:
yes